File: app/src/main/java/com/kerimovscreations/lateandroid/workers/NotifyWorker.java
package com.kerimovscreations.lateandroid.workers;

import android.app.PendingIntent;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;

import com.kerimovscreations.lateandroid.R;
import com.kerimovscreations.lateandroid.activities.MainActivity;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

public class NotifyWorker extends Worker {

    public NotifyWorker(@NonNull Context context, @NonNull WorkerParameters params) {
        super(context, params);
    }

    @NonNull
    @Override
    public Worker.Result doWork() {
        triggerNotification();

        return Worker.Result.success();
    }

    private void triggerNotification() {
        Intent intent = new Intent(getApplicationContext(), MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        PendingIntent pendingIntent = PendingIntent.getActivity(getApplicationContext(), 0, intent, 0);

        final int DBEventID = getInputData().getInt("EVENT_ID", 1);

        int soundId = getInputData().getInt("SOUND_ID", R.raw.en_male_mins_0_left);
        Uri sound = Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" + getApplicationContext().getPackageName() + "/" + soundId);
        String title = getInputData().getString("TITLE");

        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(getApplicationContext(), "B")
                .setSmallIcon(R.drawable.ic_stat_icon)
                .setSound(null)
                .setContentTitle("LATE")
                .setContentText(title)
                .setVibrate(new long[] {0, 250, 250, 250})
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setContentIntent(pendingIntent);

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(getApplicationContext());

        try {
            Ringtone r = RingtoneManager.getRingtone(getApplicationContext(), sound);
            r.play();
        } catch (Exception e) {
            e.printStackTrace();
        }

        notificationManager.notify(DBEventID, mBuilder.build());
    }
}

File: app/src/main/java/com/kerimovscreations/lateandroid/workers/NotifyWorker.kt
package com.kerimovscreations.lateandroid.workers

import android.Manifest
import android.app.PendingIntent
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.kerimovscreations.lateandroid.R
import com.kerimovscreations.lateandroid.activities.MainActivity

class NotifyWorker(private val context: Context, params: WorkerParameters) :
    Worker(context, params) {

    override fun doWork(): Result {
        triggerNotification()
        return Result.success()
    }

    private fun triggerNotification() {
        val intent = Intent(applicationContext, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        val pendingIntent = PendingIntent.getActivity(
            applicationContext, 0, intent,
            PendingIntent.FLAG_IMMUTABLE
        )
        val eventId = inputData.getInt("EVENT_ID", 1)

        val customSoundUrl = (inputData.getString("SOUND_URL") ?: "")

        val sound = if (customSoundUrl.isNotEmpty()) {
            if (hasExternalStoragePermission(applicationContext)) {
                Uri.parse(customSoundUrl)
            } else {
                val soundId = R.raw.en_male_mins_0_left
                Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" + applicationContext.packageName + "/" + soundId)
            }
        } else {
            val soundId = inputData.getInt("SOUND_ID", R.raw.en_male_mins_0_left)
            Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" + applicationContext.packageName + "/" + soundId)
        }

        val title = inputData.getString("TITLE")
        val mBuilder = NotificationCompat.Builder(applicationContext, "B")
            .setSmallIcon(R.drawable.ic_stat_icon)
            .setColor(ContextCompat.getColor(applicationContext, R.color.colorPrimary))
            .setSound(null)
            .setContentTitle("LATE")
            .setContentText(title)
            .setVibrate(longArrayOf(0, 250, 250, 250))
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setContentIntent(pendingIntent)
        val notificationManager = NotificationManagerCompat.from(applicationContext)
        try {
            val r = RingtoneManager.getRingtone(applicationContext, sound)
            r.play()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.POST_NOTIFICATIONS
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        notificationManager.notify(eventId, mBuilder.build())
    }

    private fun hasExternalStoragePermission(context: Context): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val result = context.checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE)
            return result == PackageManager.PERMISSION_GRANTED;
        }
        return false
    }
}